File: Bitsy.VB/Parser.vb
Module Parser
    Private Property Lexer As Lexer
    Private outputbuffer As List(Of String)

    Function Parse(Lexer As Lexer) As String()
        outputbuffer = New List(Of String)()
        Parser.Lexer = Lexer
        Program()
        Return outputbuffer.ToArray
    End Function

    Private Sub Program()
        Match(TokenType.Begin)
        Setup()

        Block()

        Match(TokenType.End, False)
        Teardown()
    End Sub

    Private Sub Block()
        Do Until Lexer.Current.Type = TokenType.End OrElse Lexer.Current.Type = TokenType.Else
            Select Case Lexer.Current.Type
                Case TokenType.IfN, TokenType.IfZ, TokenType.IfP
                    [If]()
                Case TokenType.Loop
                    [Loop]()
                Case TokenType.Break
                    [Break]()
                Case TokenType.Print
                    Print()
                Case TokenType.Read
                    Read()
                Case Else
                    Assignment()
            End Select
        Loop
    End Sub

#Region "Expressions"
    Private Sub Expression()
        Term()

        While Lexer.Current.Value Like "[+-]"
            Push()
            Dim op = Lexer.Current.Type
            Match(op)
            Term()
            Pop(op)
        End While
    End Sub

    Private Sub Term()
        SignedFactor()

        While Lexer.Current.Value Like "[*/%]"
            Push()
            Dim op = Lexer.Current.Type
            Match(op)
            Factor()
            Pop(op)
        End While
    End Sub

    Private Sub SignedFactor()
        Dim op = TokenType.Plus
        If Lexer.Current.Value Like "[+-]" Then
            op = Lexer.Current.Type
            Match(op)
        End If

        Factor()

        If op = TokenType.Dash Then
            Emit("Register = -Register")
        End If
    End Sub

    Private Sub Factor()
        If Lexer.Current.Type = TokenType.IntLiteral Then
            Dim int = Match(TokenType.IntLiteral)
            Emit($"Register = {int}")
        ElseIf Lexer.Current.Type = TokenType.Variable Then
            Dim varname = Match(TokenType.Variable)
            Emit($"Register = Variable(""{varname}"")")
        Else
            Match(TokenType.LeftParen)
            Expression()
            Match(TokenType.RightParen)
        End If
    End Sub

    Private Sub Push()
        Emit("Stack.Push(Register)")
    End Sub

    Private Sub Pop(Optional optype As TokenType? = Nothing)
        If optype Is Nothing Then
            Emit("Register = Stack.Pop()")
            Return
        End If
        Dim op$ = ""
        Select Case optype.Value
            Case TokenType.Plus
                op = "+"
            Case TokenType.Dash
                op = "-"
            Case TokenType.Star
                op = "*"
            Case TokenType.Slash
                op = "\"
            Case TokenType.Percent
                op = "Mod"
        End Select
        Emit($"Register = Stack.Pop() {op} Register")
    End Sub
#End Region

    Private Sub [If]()
        Dim condtype = Lexer.Current.Type
        Match(condtype)
        Expression()

        Dim comp = ""
        Select Case condtype
            Case TokenType.IfN
                comp = "<"
            Case TokenType.IfZ
                comp = "="
            Case TokenType.IfP
                comp = ">"
        End Select

        Emit($"If Register {comp} 0 Then")

        Block()

        If Lexer.Current.Type = TokenType.Else Then
            Match(TokenType.Else)
            Emit("Else")
            Block()
        End If

        Match(TokenType.End)
        Emit("End If")
    End Sub

    Private Sub [Loop]()
        Match(TokenType.Loop)
        Emit("Do")

        Block()

        Match(TokenType.End)
        Emit("Loop")
    End Sub

    Private Sub [Break]()
        Match(TokenType.Break)
        Emit("Exit Do")
    End Sub

    Private Sub Print()
        Match(TokenType.Print)
        Expression()
        Emit("Console.WriteLine(Register)")
    End Sub

    Private Sub Read()
        Match(TokenType.Read)
        Dim varname = Match(TokenType.Variable)
        Emit($"Variable(""{varname}"") = ReadIn()")
    End Sub

    Private Sub Assignment()
        Dim varname = Match(TokenType.Variable)
        Match(TokenType.Equals)
        Expression()
        Emit($"Variable(""{varname}"") = Register")
    End Sub

    Private Function Match(Type As TokenType, Optional Advance As Boolean = True) As String
        If Not Lexer.Current.Type = Type Then
            Throw New ArgumentException($"Error: Expected {Type} but received {Lexer.Current.Type}.")
        End If

        Dim retval$ = Lexer.Current.Value
        If Advance Then Lexer.Advance()
        Return retval
    End Function

    Private Sub Emit(output$)
        outputbuffer.Add(vbTab & output)
    End Sub

    Sub Setup()
        Emit("Imports System : Imports System.Collections.Generic
Module Program
    Class VariableDictionary
        Inherits Dictionary(Of String, Integer)

        Default Public Shadows Property Subscript(key As String) As Integer
            Get
                If ContainsKey(key) Then
                    Return Item(key)
                Else
                    Return 0
                End If
            End Get
            Set(val As Integer)
                Item(key) = val
            End Set
        End Property
    End Class

    Dim Register As Integer = 0
    ReadOnly Variable As New VariableDictionary()
    Dim Stack As New Stack(Of Integer)
    Function ReadIn() As Integer
        Dim input = Console.ReadLine()
        Try
            Return Convert.ToInt32(input)
        Catch ex As FormatException
            Return 0
        End Try
    End Function

    Sub Main()

    'BEGIN USER GENERATED CODE")
    End Sub

    Sub Teardown()
        Emit("  'END USER GENERATED CODE

    Console.WriteLine(""Press any key to continue..."")
    Console.ReadKey()
End Sub
End Module")
    End Sub
End Module

File: Bitsy.CS/Parser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Bitsy.CS
{
    static class Parser
    {
        private static Lexer Lexer { get; set; }
        private static List<string> outputbuffer;

        public static string[] Parse(Lexer lexer)
        {
            outputbuffer = new List<string>();
            Lexer = lexer;
            Program();
            return outputbuffer.ToArray();
        }

        private static void Program()
        {
            Match(TokenType.Begin);
            Setup();

            Block();

            Match(TokenType.End, false);
            Teardown();
        }

        private static void Block()
        {
            while (Lexer.Current.Type != TokenType.End && Lexer.Current.Type != TokenType.Else)
            {
                switch (Lexer.Current.Type)
                {
                    case TokenType.IfN:
                    case TokenType.IfZ:
                    case TokenType.IfP:
                        If();
                        break;

                    case TokenType.Loop:
                        Loop();
                        break;

                    case TokenType.Break:
                        Break();
                        break;

                    case TokenType.Print:
                        Print();
                        break;

                    case TokenType.Read:
                        Read();
                        break;

                    default:
                        Assignment();
                        break;
                }
            }
        }

        #region Expressions
        private static void Expression()
        {
            Term();

            while (Lexer.Current.Value == "+" || Lexer.Current.Value == "-")
            {
                Push();
                TokenType op = Lexer.Current.Type;
                Match(op);
                Term();
                Pop(op);
            }
        }

        private static void Term()
        {
            SignedFactor();

            while (Lexer.Current.Value == "*" || Lexer.Current.Value == "/" || Lexer.Current.Value == "%")
            {
                Push();
                TokenType op = Lexer.Current.Type;
                Match(op);
                Factor();
                Pop(op);
            }
        }

        private static void SignedFactor()
        {
            TokenType op = TokenType.Plus;
            if (Lexer.Current.Value == "+" || Lexer.Current.Value == "-")
            {
                op = Lexer.Current.Type;
                Match(op);
            }

            Factor();

            if (op == TokenType.Dash)
                Emit("register = -register");
        }

        private static void Factor()
        {
            if (Lexer.Current.Type == TokenType.IntLiteral)
            {
                string integer = Match(TokenType.IntLiteral);
                Emit($"register = {integer}");
            }
            else if (Lexer.Current.Type == TokenType.Variable)
            {
                string varname = Match(TokenType.Variable);
                Emit($"register = variable[\"{varname}\"]");
            }
            else
            {
                Match(TokenType.LeftParen);
                Expression();
                Match(TokenType.RightParen);
            }
        }

        private static void Push()
        {
            Emit("stack.Push(register)");
        }

        private static void Pop(TokenType? optype = null)
        {
            if (optype == null)
            {
                Emit("register = stack.Pop()");
                return;
            }
            char op;
            switch (optype.Value)
            {
                case TokenType.Plus:
                    op = '+';
                    break;
                case TokenType.Dash:
                    op = '-';
                    break;
                case TokenType.Star:
                    op = '*';
                    break;
                case TokenType.Slash:
                    op = '/';
                    break;
                case TokenType.Percent:
                    op = '%';
                    break;
                default:
                    throw new NotSupportedException("Unreachable code hit.");
            }
            Emit($"register = stack.Pop() {op} register");
        }
        #endregion

        private static void If()
        {
            TokenType condtype = Lexer.Current.Type;
            Match(condtype);
            Expression();

            string comp = "";
            switch (condtype)
            {
                case TokenType.IfN:
                    comp = "<";
                    break;
                case TokenType.IfZ:
                    comp = "==";
                    break;
                case TokenType.IfP:
                    comp = ">";
                    break;
            }

            Emit($"if (register {comp} 0)\n\t{{", false);

            Block();

            if (Lexer.Current.Type == TokenType.Else)
            {
                Match(TokenType.Else);
                Emit("}\n\telse\n\t{", false);
                Block();
            }

            Match(TokenType.End);
            Emit("}", false);
        }

        private static void Loop()
        {
            Match(TokenType.Loop);
            Emit("while (true)\n\t{", false);

            Block();

            Match(TokenType.End);
            Emit("}", false);
        }

        private static void Break()
        {
            Match(TokenType.Break);
            Emit("break");
        }

        private static void Print()
        {
            Match(TokenType.Print);
            Expression();
            Emit("Console.WriteLine(register)");
        }

        private static void Read()
        {
            Match(TokenType.Read);
            string varname = Match(TokenType.Variable);
            Emit($"variable[\"{varname}\"] = ReadIn()");
        }

        private static void Assignment()
        {
            string varname = Match(TokenType.Variable);
            Match(TokenType.Equals);
            Expression();
            Emit($"variable[\"{varname}\"] = register");
        }

        private static string Match(TokenType type, bool advance = true)
        {
            if (Lexer.Current.Type != type)
                throw new ArgumentException($"Error: Expected {type} but received {Lexer.Current.Type}.");
            string retval = Lexer.Current.Value;
            if (advance)
                Lexer.Advance();
            return retval;
        }

        private static void Emit(string output, bool semi = true) => outputbuffer.Add('\t' + output + (semi ? ";" : string.Empty));

        private static void Setup() => Emit("" +
            @"using System; using System.Collections.Generic;
namespace BitsyExec {
class VariableDictionary : Dictionary<string, int> {public new int this[string key] { get { return ContainsKey(key) ? base[key] : 0; } set { base[key] = value;} }}
class Program
{
private static int register = 0;
static readonly VariableDictionary variable = new VariableDictionary();
static readonly Stack<int> stack = new Stack<int>();
static int ReadIn() { try { return Convert.ToInt32(Console.ReadLine()); } catch { return 0; }}
public static void Main()
{

    // BEGIN USER GENERATED CODE", false);
        private static void Teardown() => Emit("" +
            @"  // END USER GENERATED CODE

}}}", false);
    }
}